services/src/main/java/org/keycloak/vault/DefaultVaultTranscriber.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.vault;
+
+import java.lang.ref.WeakReference;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Default {@link VaultTranscriber} implementation that uses the configured {@link VaultProvider} to obtain raw secrets
+ * and convert them into other types. By default, the {@link VaultProvider} provides raw secrets through a {@link ByteBuffer}.
+ * This class offers methods to convert the raw secrets into other types (such as {@link VaultCharSecret} or {@link WeakReference<String>}).
+ *
+ * @see VaultRawSecret
+ * @see VaultCharSecret
+ *
+ * @author <a href="mailto:sguilhen@redhat.com">Stefan Guilhen</a>
+ */
+public class DefaultVaultTranscriber implements VaultTranscriber {
+
+    private static final Pattern pattern = Pattern.compile("^\\$\\{vault\\.(.+)}$");
+
+    private final VaultProvider provider;
+
+    public DefaultVaultTranscriber(final VaultProvider provider) {
+        this.provider = provider;
+    }
+
+    @Override
+    public VaultRawSecret getRawSecret(final String value) {
```

Comment:
I was wondering about a situation when isVaultExpression returns true (value is in the form of ${vault.xyz}) and we are unable to transcribe the value from the vault. Should we somehow notify users this situation happen? I think in the majority of cases this will be caused by a wrongly configured vault (maybe a typo in vault id, etc.). The result of this would be for example _unable to connect to the SMTP server (wrong password)_, which is actually not the root cause. Maybe we can somehow let users know that this is actually caused by vault configuration. Does it make sense?

Reply:
Not here. If the value is not found, `VaultProvider` should write it to the log.

Reply:
I can't see this log message at the moment in PlainTextVaultProvider. Should we add it as part of this PR. Or add it later?

Reply:
Let's add it later.